File: internal/hir/verify.go
package hir

import (
	"fmt"
	"strings"
)

type VerificationError struct {
	Function *FunctionID
	Message  string
}

type VerificationErrors []VerificationError

func (e VerificationErrors) Error() string {
	parts := make([]string, 0, len(e))
	for _, item := range e {
		if item.Function == nil {
			parts = append(parts, item.Message)
			continue
		}
		parts = append(parts, fmt.Sprintf("function f%d: %s", *item.Function, item.Message))
	}
	return strings.Join(parts, "; ")
}

func (m Module) Verify() error {
	var errors VerificationErrors
	errors = append(errors, m.verifyTypes()...)
	errors = append(errors, m.verifyShapes()...)
	errors = append(errors, m.verifyFunctions()...)
	if len(errors) == 0 {
		return nil
	}
	return errors
}
func (m Module) verifyTypes() VerificationErrors {
	var errors VerificationErrors
	for index, semantic := range m.Types {
		check := func(target TypeID) {
			if int(target) >= len(m.Types) {
				errors = append(errors, VerificationError{Message: fmt.Sprintf("type t%d references invalid type t%d", index, target)})
			}
		}
		switch semantic.Kind {
		case TypeArray:
			check(semantic.Element)
		case TypeUnion:
			for _, member := range semantic.Members {
				check(member)
			}
		case TypeFunction:
			for _, param := range semantic.Params {
				check(param)
			}
			check(semantic.ReturnType)
		}
	}
	return errors
}

func (m Module) verifyShapes() VerificationErrors {
	var errors VerificationErrors
	seen := map[ShapeID]struct{}{}
	for _, shape := range m.Shapes {
		if _, ok := seen[shape.ID]; ok {
			errors = append(errors, VerificationError{Message: fmt.Sprintf("duplicate shape s%d", shape.ID)})
		}
		seen[shape.ID] = struct{}{}
		for _, field := range shape.Fields {
			if int(field.SemanticType) >= len(m.Types) {
				errors = append(errors, VerificationError{Message: fmt.Sprintf("shape s%d field %s references invalid type t%d", shape.ID, field.Name, field.SemanticType)})
			}
		}
	}
	for i, typ := range m.Types {
		if typ.Kind == TypeObject {
			if _, ok := seen[typ.Shape]; !ok {
				errors = append(errors, VerificationError{Message: fmt.Sprintf("object type t%d references unknown shape s%d", i, typ.Shape)})
			}
		}
	}
	return errors
}

func (m Module) verifyFunctions() VerificationErrors {
	var errors VerificationErrors
	functionIDs := make(map[FunctionID]struct{}, len(m.Functions))
	for _, function := range m.Functions {
		if _, exists := functionIDs[function.ID]; exists {
			errors = append(errors, VerificationError{Message: fmt.Sprintf("duplicate function id f%d", function.ID)})
		}
		functionIDs[function.ID] = struct{}{}
	}
	if m.Entry != nil {
		if _, exists := functionIDs[*m.Entry]; !exists {
			errors = append(errors, VerificationError{Message: fmt.Sprintf("entry references unknown function f%d", *m.Entry)})
		}
	}
	for i := range m.Functions {
		errors = append(errors, m.verifyFunction(&m.Functions[i], functionIDs)...)
	}
	return errors
}
func (m Module) verifyFunction(function *Function, functionIDs map[FunctionID]struct{}) VerificationErrors {
	var errors VerificationErrors
	add := func(message string) {
		id := function.ID
		errors = append(errors, VerificationError{Function: &id, Message: message})
	}
	checkType := func(typeID TypeID) {
		if int(typeID) >= len(m.Types) {
			add(fmt.Sprintf("references invalid type t%d", typeID))
		}
	}

	checkType(function.ReturnType)
	blockIDs := make(map[BlockID]struct{}, len(function.Blocks))
	for _, block := range function.Blocks {
		if _, exists := blockIDs[block.ID]; exists {
			add(fmt.Sprintf("duplicate block id b%d", block.ID))
		}
		blockIDs[block.ID] = struct{}{}
	}
	if _, exists := blockIDs[function.Entry]; !exists {
		add(fmt.Sprintf("missing entry block b%d", function.Entry))
	}

	values := make(map[ValueID]struct{})
	for _, param := range function.Params {
		checkType(param.SemanticType)
		if _, exists := values[param.Value]; exists {
			add(fmt.Sprintf("duplicate value id v%d", param.Value))
		}
		values[param.Value] = struct{}{}
	}
	for _, block := range function.Blocks {
		for _, instruction := range block.Instructions {
			checkType(instruction.SemanticType)
			if _, exists := values[instruction.Result]; exists {
				add(fmt.Sprintf("duplicate value id v%d", instruction.Result))
			}
			values[instruction.Result] = struct{}{}
		}
	}

	checkValue := func(value ValueID) {
		if _, exists := values[value]; !exists {
			add(fmt.Sprintf("references unknown value v%d", value))
		}
	}
	checkBlock := func(block BlockID) {
		if _, exists := blockIDs[block]; !exists {
			add(fmt.Sprintf("references unknown block b%d", block))
		}
	}

	for _, block := range function.Blocks {
		for _, instruction := range block.Instructions {
			switch op := instruction.Op.(type) {
			case ConstOp:
			case UnaryExpr:
				checkValue(op.Operand)
			case BinaryExpr:
				checkValue(op.Left)
				checkValue(op.Right)
			case CallOp:
				if _, exists := functionIDs[op.Callee]; !exists {
					add(fmt.Sprintf("calls unknown function f%d", op.Callee))
				}
				for _, arg := range op.Args {
					checkValue(arg)
				}
			case IntrinsicCallOp:
				if op.Intrinsic == IntrinsicInvalid {
					add(fmt.Sprintf("instruction v%d has invalid intrinsic", instruction.Result))
				}
				for _, arg := range op.Args {
					checkValue(arg)
				}
			case PhiOp:
				if len(op.Incoming) == 0 {
					add(fmt.Sprintf("phi v%d has no incoming values", instruction.Result))
				}
				for _, incoming := range op.Incoming {
					checkBlock(incoming.Block)
					checkValue(incoming.Value)
				}
			case ArrayNewOp:
				for _, element := range op.Elements {
					checkValue(element)
				}
			case ArrayLengthOp:
				checkValue(op.Array)
			case ArrayGetOp:
				checkValue(op.Array)
				checkValue(op.Index)
			case ObjectNewOp:
				if int(op.Shape) >= len(m.Shapes) {
					add(fmt.Sprintf("object allocation references unknown shape s%d", op.Shape))
				}
				for _, field := range op.Fields {
					checkValue(field)
				}
			case FieldGetOp:
				checkValue(op.Object)
				if int(op.Shape) >= len(m.Shapes) {
					add(fmt.Sprintf("field access references unknown shape s%d", op.Shape))
				} else if int(op.Field) >= len(m.Shapes[op.Shape].Fields) {
					add(fmt.Sprintf("field access s%d.%d is out of range", op.Shape, op.Field))
				}
			case ClosureNewOp:
				if _, exists := functionIDs[op.Callee]; !exists {
					add(fmt.Sprintf("closure references unknown function f%d", op.Callee))
				}
				for _, capture := range op.Captures {
					checkValue(capture)
				}
			case ClosureCallOp:
				checkValue(op.Closure)
				for _, arg := range op.Args {
					checkValue(arg)
				}
			case nil:
				add(fmt.Sprintf("instruction v%d has nil operation", instruction.Result))
			}
		}
		switch term := block.Terminator.(type) {
		case ReturnTerm:
			if term.Value != nil {
				checkValue(*term.Value)
			}
		case JumpTerm:
			checkBlock(term.Target)
		case BranchTerm:
			checkValue(term.Condition)
			checkBlock(term.Then)
			checkBlock(term.Else)
		case nil:
			add(fmt.Sprintf("block b%d has nil terminator", block.ID))
		}
	}
	return errors
}

File: crates/tsnative-ir/src/verify.rs
use std::collections::HashSet;

use thiserror::Error;

use crate::{
    BlockId, FunctionId, HirFunction, HirInstructionKind, HirModule, HirTerminator, TypeId, ValueId,
};

#[derive(Debug, Clone, PartialEq, Eq, Error)]
pub enum VerifyError {
    #[error("duplicate function id {0:?}")]
    DuplicateFunction(FunctionId),
    #[error("function {function:?} has duplicate block id {block:?}")]
    DuplicateBlock {
        function: FunctionId,
        block: BlockId,
    },
    #[error("function {function:?} has duplicate value id {value:?}")]
    DuplicateValue {
        function: FunctionId,
        value: ValueId,
    },
    #[error("function {function:?} references invalid type id {type_id:?}")]
    InvalidType {
        function: FunctionId,
        type_id: TypeId,
    },
    #[error("function {function:?} has missing entry block {block:?}")]
    MissingEntryBlock {
        function: FunctionId,
        block: BlockId,
    },
    #[error("function {function:?} references unknown block {block:?}")]
    UnknownBlock {
        function: FunctionId,
        block: BlockId,
    },
    #[error("function {function:?} references unknown value {value:?}")]
    UnknownValue {
        function: FunctionId,
        value: ValueId,
    },
    #[error("function {function:?} calls unknown function {callee:?}")]
    UnknownFunction {
        function: FunctionId,
        callee: FunctionId,
    },
}

impl HirModule {
    pub fn verify(&self) -> Result<(), Vec<VerifyError>> {
        verify_module(self)
    }
}

pub fn verify_module(module: &HirModule) -> Result<(), Vec<VerifyError>> {
    let mut errors = Vec::new();
    let mut function_ids = HashSet::new();

    for function in &module.functions {
        if !function_ids.insert(function.id) {
            errors.push(VerifyError::DuplicateFunction(function.id));
        }
    }

    for function in &module.functions {
        verify_function(module, function, &function_ids, &mut errors);
    }

    if errors.is_empty() {
        Ok(())
    } else {
        Err(errors)
    }
}

fn verify_function(
    module: &HirModule,
    function: &HirFunction,
    function_ids: &HashSet<FunctionId>,
    errors: &mut Vec<VerifyError>,
) {
    verify_type(module, function.id, function.return_type, errors);

    let mut block_ids = HashSet::new();
    for block in &function.blocks {
        if !block_ids.insert(block.id) {
            errors.push(VerifyError::DuplicateBlock {
                function: function.id,
                block: block.id,
            });
        }
    }

    if !block_ids.contains(&function.entry) {
        errors.push(VerifyError::MissingEntryBlock {
            function: function.id,
            block: function.entry,
        });
    }

    let mut values = HashSet::new();
    for param in &function.params {
        verify_type(module, function.id, param.semantic_type, errors);
        insert_value(function.id, param.value, &mut values, errors);
    }

    for block in &function.blocks {
        for instruction in &block.instructions {
            verify_type(module, function.id, instruction.semantic_type, errors);
            insert_value(function.id, instruction.result, &mut values, errors);
        }
    }

    for block in &function.blocks {
        for instruction in &block.instructions {
            verify_instruction(
                function.id,
                &instruction.kind,
                &values,
                function_ids,
                errors,
            );
        }
        verify_terminator(function.id, &block.terminator, &values, &block_ids, errors);
    }
}

fn verify_type(
    module: &HirModule,
    function: FunctionId,
    type_id: TypeId,
    errors: &mut Vec<VerifyError>,
) {
    if type_id.0 as usize >= module.types.len() {
        errors.push(VerifyError::InvalidType { function, type_id });
    }
}

fn insert_value(
    function: FunctionId,
    value: ValueId,
    values: &mut HashSet<ValueId>,
    errors: &mut Vec<VerifyError>,
) {
    if !values.insert(value) {
        errors.push(VerifyError::DuplicateValue { function, value });
    }
}

fn verify_value(
    function: FunctionId,
    value: ValueId,
    values: &HashSet<ValueId>,
    errors: &mut Vec<VerifyError>,
) {
    if !values.contains(&value) {
        errors.push(VerifyError::UnknownValue { function, value });
    }
}

fn verify_instruction(
    function: FunctionId,
    kind: &HirInstructionKind,
    values: &HashSet<ValueId>,
    functions: &HashSet<FunctionId>,
    errors: &mut Vec<VerifyError>,
) {
    match kind {
        HirInstructionKind::Const(_) => {}
        HirInstructionKind::Unary { operand, .. } => {
            verify_value(function, *operand, values, errors);
        }
        HirInstructionKind::Binary { left, right, .. } => {
            verify_value(function, *left, values, errors);
            verify_value(function, *right, values, errors);
        }
        HirInstructionKind::Call { callee, args } => {
            if !functions.contains(callee) {
                errors.push(VerifyError::UnknownFunction {
                    function,
                    callee: *callee,
                });
            }
            for arg in args {
                verify_value(function, *arg, values, errors);
            }
        }
    }
}

fn verify_terminator(
    function: FunctionId,
    terminator: &HirTerminator,
    values: &HashSet<ValueId>,
    blocks: &HashSet<BlockId>,
    errors: &mut Vec<VerifyError>,
) {
    match terminator {
        HirTerminator::Return(value) => {
            if let Some(value) = value {
                verify_value(function, *value, values, errors);
            }
        }
        HirTerminator::Jump(target) => {
            verify_block(function, *target, blocks, errors);
        }
        HirTerminator::Branch {
            condition,
            then_block,
            else_block,
        } => {
            verify_value(function, *condition, values, errors);
            verify_block(function, *then_block, blocks, errors);
            verify_block(function, *else_block, blocks, errors);
        }
    }
}

fn verify_block(
    function: FunctionId,
    block: BlockId,
    blocks: &HashSet<BlockId>,
    errors: &mut Vec<VerifyError>,
) {
    if !blocks.contains(&block) {
        errors.push(VerifyError::UnknownBlock { function, block });
    }
}
